### PR TITLE
[giveth-xdai-balance] added two new strategies for GIveth

### PR DESCRIPTION
**Pull request by pkretzschmar**
Added two new streategies:

- giveth-xdai-balance
- giveth-balancer-balance

### CODE REVIEW

File: src/strategies/giveth-xdai-balance/index.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  return amountLP.mul(givBalance).div(totalLP);
+};
+
+export async function strategy(
+  space,
+  network,
+  provider,
+  addresses,
+  options,
+  snapshot
```
Comment (ChaituVR) [inline src/strategies/giveth-xdai-balance/index.ts:63]: Hi @pkretzschmar you may have to request data from subgraph based on snapshot or its timestamp, else the voting power will keep changing in future

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/strategies/giveth-xdai-balance/index.ts (hunk 1 above)
Comment (pkretzschmar) [reply src/strategies/giveth-xdai-balance/index.ts:79]: Done!

File: src/strategies/giveth-balancer-balance/index.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  }
+  const balData = await subgraphRequest(BALANCER_SUBGRAPH_API, poolParams);
+  const balFormatedData = formatReserveBalance(balData, options.decimals);
+
+  const data = await Promise.all(
+    addresses.map((address) => {
+      params.balance.__args.id = address.toLowerCase();
+      return subgraphRequest(GIVETH_SUBGRAPH_API, params);
+    })
+  );
```
Comment (ChaituVR) [inline src/strategies/giveth-balancer-balance/index.ts:76]: Hi @pkretzschmar Calling the subgraph per one address won't scale for proposals having large voters, Is there a way to request data for all addresses at once?

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/strategies/giveth-balancer-balance/index.ts (hunk 2 above)
Comment (pkretzschmar) [reply src/strategies/giveth-balancer-balance/index.ts]: I found a way, just pushed it!

File: src/strategies/giveth-balancer-balance/index.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      balancerLpStaked
+    } = addressBalance;
+    const totalGIV = BigNumber.from(balance).add(givStaked);
+
+    const balGIV = calcGivAmount(
+      BigNumber.from(balancerLp).add(balancerLpStaked),
+      balFormatedData.totalShares,
+      balFormatedData.balance
+    );
+    score[id] = parseFloat(formatUnits(totalGIV.add(balGIV), options.decimals));
```
Comment (ChaituVR) [inline src/strategies/giveth-balancer-balance/index.ts:99]: Looks good @pkretzschmar Just one small fix needed, looks like all the addresses returned from subgraph are lowercase, we need to return checksum addresses, you can use `getAdress` for that, example
https://github.com/snapshot-labs/snapshot-strategies/blob/master/src/strategies/synthetix-quadratic/index.ts#L124

We need to do same thing for `giveth-xdai-balance` strategy too

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/strategies/giveth-balancer-balance/index.ts (hunk 3 above)
Comment (pkretzschmar) [reply src/strategies/giveth-balancer-balance/index.ts]: Done! Thanks for your feedbacks!

File: src/strategies/giveth-xdai-balance/index.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+  params.balances.__args.where.id_in = addresses.map((address) =>
+    address.toLowerCase()
+  );
+
+  const data = await subgraphRequest(GIVETH_SUBGRAPH_API, params);
+  const dataBalances = data.balances;
+
+  const score = {};
+  console.log(dataBalances);
```
Comment (ChaituVR) [inline src/strategies/giveth-xdai-balance/index.ts:114]:
```suggestion
```

**Comment by ChaituVR** [comment]
image

Fail when passing latest snapshot block number

**Review by ChaituVR — APPROVED** [review]

**Pull request merged**